File: 2022/python/day02.py
import os
from typing import Tuple

scores = {
    "X": 1,
    "Y": 2,
    "Z": 3,
    "A": 1,
    "B": 2,
    "C": 3,
}

stra_win = ["B", "C", "A"]
stra_draw = ["A", "B", "C"]
stra_lose = ["C", "A", "B"]


def compute_outcome1(left: str, right: str) -> int:
    delta = scores[right] - scores[left]
    if (delta == 1) or (delta == -2):
        return scores[right] + 6
    elif delta == 0:
        return scores[right] + 3
    else:
        return scores[right]


def compute_outcome2(left: str, right: str) -> int:
    if right == "X":
        return scores[stra_lose[scores[left] - 1]]
    elif right == "Y":
        return scores[stra_draw[scores[left] - 1]] + 3
    else:
        return scores[stra_win[scores[left] - 1]] + 6


def parse_line(line: str) -> Tuple[str, str]:
    inputs = line.rstrip("\n").split(" ")
    return (inputs[0], inputs[1])


def main():
    with open("../input/day2.txt", "r") as handle:
        lines = handle.readlines()

    sum_score1 = 0
    sum_score2 = 0
    for line in lines:
        if len(line) > 1:
            numbers = parse_line(line)
            sum_score1 += compute_outcome1(*numbers)
            sum_score2 += compute_outcome2(*numbers)

    print(f"day2-part1: {sum_score1}")
    print(f"day2-part2: {sum_score2}")


if __name__ == "__main__":
    main()

File: 2022/rust/src/day02.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("../input/day2.txt").unwrap();
    let mut sum_score1: u32 = 0;
    let mut sum_score2: u32 = 0;

    for line in input.lines() {
        if line.len() == 0 {
            continue;
        } else {
            let (left, right) = parse_line(line);
            sum_score1 += compute_outcome1(&left, &right);
            sum_score2 += compute_outcome2(&left, &right);
        }
    }

    println!("day2-part1: {}", sum_score1);
    println!("day2-part2: {}", sum_score2);
}

fn parse_line(line: &str) -> (String, String) {
    let output: Vec<String> = line.trim().split(" ").map(|m| m.to_string()).collect();

    (
        output.get(0).unwrap().to_owned(),
        output.get(1).unwrap().to_owned(),
    )
}

fn compute_outcome1(left: &String, right: &String) -> u32 {
    let val_left = match left.as_str() {
        "A" => 1,
        "B" => 2,
        "C" => 3,
        _ => 4,
    };
    let val_right = match right.as_str() {
        "X" => 1,
        "Y" => 2,
        "Z" => 3,
        _ => 4,
    };

    let delta: i32 = val_right as i32 - val_left as i32;
    if (delta == 1) || (delta == -2) {
        val_right + 6
    } else if delta == 0 {
        val_right + 3
    } else {
        val_right
    }
}
fn compute_outcome2(left: &String, right: &String) -> u32 {
    let val_left = match left.as_str() {
        "A" => 1,
        "B" => 2,
        "C" => 3,
        _ => 4,
    };
    let retval = match right.as_str() {
        "X" => {
            if val_left == 1 {
                3
            } else {
                val_left - 1
            }
        }
        "Y" => 3 + val_left,
        "Z" => {
            if val_left == 3 {
                6 + 1
            } else {
                6 + val_left + 1
            }
        }
        _ => 4,
    };

    retval
}

#[cfg(test)]
mod test {
    use super::*;

    const INPUT: &str = r"A Y
B X
C Z";

    #[test]
    fn part2_test() {
        let mut result = 0;
        for line in INPUT.lines() {
            let (left, right) = parse_line(line);
            result += compute_outcome2(&left, &right);
        }
        assert_eq!(result, 12);
    }
}
